util/pingpong.c

```diff
@@ -0,0 +1,2229 @@
+
+	sizes_cnt = generate_test_sizes(&ct->opts, ct->fi->ep_attr->max_msg_size, &sizes, PP_SIZE_MAX_POWER_TWO);
+
+	PP_DEBUG("Count of sizes to test : %d\n", sizes_cnt);
+
+	for (i = 0; i < sizes_cnt; i++) {
+		ct->opts.transfer_size = sizes[i];
+		if (ct->opts.transfer_size > ct->fi->ep_attr->max_msg_size) {
+			PP_DEBUG("Transfer size too high for endpoint : %d\n", ct->opts.transfer_size);
+			continue;
+		}
+		init_test(ct, &(ct->opts), ct->test_name, sizeof(ct->test_name));
+		ret = pingpong(ct);
+		if (ret)
+			return ret;
+	}
+
+	free(sizes);
+
+	return pp_finalize(ct);
+}
+
+static int run_pingpong_msg(struct ct_pingpong *ct)
+{
+	int i, ret, sizes_cnt;
+	int *sizes = NULL;
+
+	PP_DEBUG("Selected endpoint : MSG\n");
+
+	ret = pp_ctrl_init(ct);
+	if (ret) {
+		return ret;
+	}
+
+	ret = pp_ctrl_txrx_data_port(ct);
+	if (ret) {
+		return ret;
+	}
+
+	if (!ct->opts.dst_addr) {
+		ret = pp_start_server(ct);
+		if (ret)
+			return ret;
+	}
+
+	ret = ct->opts.dst_addr ? pp_client_connect(ct) : pp_server_connect(ct);
+	if (ret) {
+		return ret;
+	}
+
+	pp_banner_fabric_info(ct);
+
+	sizes_cnt = generate_test_sizes(&ct->opts, ct->fi->ep_attr->max_msg_size, &sizes, PP_SIZE_MAX_POWER_TWO);
+
+	PP_DEBUG("Count of sizes to test : %d\n", sizes_cnt);
+
+	for (i = 0; i < sizes_cnt; i++) {
+		ct->opts.transfer_size = sizes[i];
+		if (ct->opts.transfer_size > ct->fi->ep_attr->max_msg_size) {
+			PP_DEBUG("Transfer size too high for endpoint : %d\n", ct->opts.transfer_size);
+			continue;
+		}
+		init_test(ct, &(ct->opts), ct->test_name, sizeof(ct->test_name));
+		ret = pingpong(ct);
+		if (ret)
+			goto out;
+	}
+
+	free(sizes);
+
+	ret = pp_finalize(ct);
+out:
+	fi_shutdown(ct->ep, 0);
+	return ret;
+}
+
+int main(int argc, char **argv)
+{
+	int ret, op;
+
+	ret = EXIT_SUCCESS;
+
+	struct ct_pingpong ct;
+
+	pp_init_ct_pingpong(&ct);
+	ct.opts = (struct pp_opts) {
+		.options = 0,
+		.iterations = 1000,
+		.transfer_size = 1024,
+		.sizes_enabled = PP_DEFAULT_SIZE,
+		.argc = argc, .argv = argv
+	};
+
+	ct.hints = fi_allocinfo();
+	if (!ct.hints)
+		return EXIT_FAILURE;
+
+	while ((op = getopt(argc, argv, "hd" "b:p:s:n:f:e:I:S:v")) != -1) {
+		switch (op) {
+		default:
+			pp_parse_opts(&ct, op, optarg);
+			break;
+		case 'd':
+			pp_debug = 1;
+			break;
+		case '?':
+		case 'h':
+			pp_pingpong_usage(argv[0], "Ping pong client and server.");
+			return EXIT_FAILURE;
+		}
+	}
+
+	if (optind < argc)
+		ct.opts.dst_addr = argv[optind];
+
+	if (!ct.hints->ep_attr->type || ct.hints->ep_attr->type == FI_EP_UNSPEC) {
+		ct.hints->ep_attr->type = FI_EP_DGRAM;
+	}
+
+	pp_banner_options(&ct);
+
+	switch(ct.hints->ep_attr->type) {
+	case FI_EP_DGRAM:
+		if (ct.opts.options & PP_OPT_SIZE)
+			ct.hints->ep_attr->max_msg_size = ct.opts.transfer_size;
+		ct.hints->caps = FI_MSG;
+		ct.hints->mode = FI_LOCAL_MR;
+		ret = run_pingpong_dgram(&ct);
+		break;
+	case FI_EP_RDM:
+		ct.hints->caps = FI_MSG;
+		ct.hints->mode = FI_CONTEXT | FI_LOCAL_MR;
```

Comment:
Yeah, it makes sense to add FI_CONTEXT since many providers require that. But, I was wondering why it is added 'only' for RDM ep type. Since, we are reusing pingpong() for all ep types, cant we add it for other types too?